klog/klog.c

```diff
@@ -0,0 +1,327 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Kernel log buffer
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include HAL
+
+#include "../include/ioctl.h"
+
+#include "../posix/posix.h"
+
+#include "klog.h"
+#include "../proc/threads.h"
+#include "../proc/msg.h"
+#include "../proc/ports.h"
+
+
+#ifndef KLOG_BUFSZ
+#define KLOG_BUFSZ (2 * SIZE_PAGE)
+#endif
+
+
+#define TCGETS 0x405c7401
+
+
+typedef struct klog_reader {
+	msg_t msg;
+	offs_t ridx;
+	unsigned long rid;
+	unsigned pid;
+	unsigned nonblocking;
+	unsigned waiting;
+	struct klog_reader *prev, *next;
+} klog_reader_t;
+
+
+static struct {
+	char buf[KLOG_BUFSZ];
+	u32 port;
+	offs_t head;
+	offs_t tail;
+	lock_t lock;
+	klog_reader_t *readers;
+} klog_common;
+
+
+static int _fifo_empty(void)
+{
+	return klog_common.tail == klog_common.head;
+}
+
+
+static int _fifo_full(void)
+{
+	return klog_common.tail - klog_common.head == KLOG_BUFSZ;
+}
+
+
+static char _fifo_pop(void)
+{
+	return klog_common.buf[klog_common.head++ % KLOG_BUFSZ];
+}
+
+
+static void _fifo_push(char c)
+{
+	klog_common.buf[klog_common.tail % KLOG_BUFSZ] = c;
+	klog_common.tail++;
+	if (_fifo_full())
+		_fifo_pop();
+}
+
+
+static char _fifo_get(offs_t off)
+{
+	return klog_common.buf[off % KLOG_BUFSZ];
+}
+
+
+void _klog_init(void)
+{
+	hal_memset(&klog_common, 0, sizeof(klog_common));
+	proc_lockInit(&klog_common.lock);
+}
+
+
+static klog_reader_t *klog_readerFind(unsigned pid)
+{
+	klog_reader_t *r = klog_common.readers;
```

Comment:
How about - use tree instead of list. This way we can use `pid` as our index.

Reply:
Yes, I was considering using a tree, but I decided that it would be a bit of an overkill, as for most of the time we would have a very limited number of readers. I can change it to an rbtree if you prefer.

Reply:
I suggest a tree mostly because I think it might simplify logic here, your call